docs/DeveloperGuide.md

```diff
@@ -160,9 +160,60 @@ Classes used by multiple components are in the `seedu.address.commons` package.
 
 This section describes some noteworthy details on how certain features are implemented.
 
-### Logic
+### Command Autocompletion, Real-Time Input Validation and Recommendation
 
-#### Add and Delete Elderly and Volunteer
+Autocompletion and Recommendation are features that can greatly enhance the user experience when using the application.
+To support this feature, we have implemented a `CommandRecommedationEngine` that aims to accurately predict the set of
+words that the user intends to type based off a preset of words. These words are the in-built commands and argument
+prefixes.
+
+In `CommandRecommendationEngine`, you can find three methods:
+
+1. `recommendCommand` -- recommends an appropriate command with respect to the user input.
+2. `autocompleteCommand` -- autocompletes the user input by replacing user input with the recommended command.
+3. `registerCommandInfo` -- registers the command, turning on command recommendation.
+
+#### Autocompletion
+
+An event handler is attached to `commandTextField` which listens for the `KEY_PRESSED` event and autocompletes the user
+input with the recommended values.
+
+#### Recommendation
+
+To predict the words, we made use of the longest prefix matching algorithm to first **identify** the related command,
+and then **verify** the arguments associated with it. The following activity diagram describes the activity flow:
+
+<img src="images/developerGuide/CommandRecommendationActivityDiagram.png"/>
+
+**Terminology**:
+
+- Full attribute command -- indicates that all arguments of a command (including optional and compulsory) have been
+  specified.
+- Complete command -- indicates that command (for example, `add_elderly`) has been fully typed. Arguments may or may not
+  have been typed.
+
+#### Input validation
+
+Our implementation strategy involves utilizing an event listener to provide immediate feedback to the user regarding the
+correctness of the attributes specified, instead of depending on the parser for this purpose. For example, if the
+`AddElderlyCommand` does not accept any arguments without a prefix but user specifies anyways, a **warning** will
+be given.
```

Comment:
```suggestion
correctness of the attributes specified, instead of depending on the parser for this purpose. For example, if the
`AddElderlyCommand` does not accept any arguments without a prefix but the user specifies anyhow, a **warning** will
be given.
```